fdbclient/PrivateKeySpace.actor.cpp

```diff
@@ -0,0 +1,205 @@
+ACTOR Future<Void> normalizeKeySelectorActor(
+    const PrivateKeyRangeBaseImpl* pkrImpl,
+    Reference<ReadYourWritesTransaction> ryw,
+    KeySelector* ks )
+{
+    ASSERT(!ks->orEqual); // should be removed before calling
+    ASSERT(ks->offset != 1); // The function is never called when KeySelector is already normalized
+    
+    state KeyRangeRef range = pkrImpl->getKeyRange();
+    state Key startKey(range.begin);
+    state Key endKey(range.end);
+
+    if (ks->offset < 1) {
+        // less than the given key
+        if (range.contains(ks->getKey()))
+            endKey = keyAfter(ks->getKey());
+    }
+    else {
+        // greater than the given key
+        if (range.contains(ks->getKey()))
+            startKey = ks->getKey();
+    }
+
+    TraceEvent("NormalizeKeySelector").
+        detail("OriginalKey", ks->getKey()).
+        detail("OriginalOffset", ks->offset).
+        detail("PrivateKeyRangeStart", range.begin).
+        detail("PrivateKeyRangeEnd", range.end);
+    
+    Standalone<RangeResultRef> result = wait(pkrImpl->getRange(ryw, KeyRangeRef(startKey, endKey)));
+    // TODO : KeySelector::setKey has byte limit according to the knobs, customize it if needed
+    if (ks->offset < 1) {
+        if (result.size() >= 1 - ks->offset) {
+            ks->setKey(KeyRef(ks->arena(), result[result.size()-(1-ks->offset)].key));
+            ks->offset = 1;
+        } else {
+            ks->setKey(KeyRef(ks->arena(), result[0].key));
+            ks->offset += result.size();
+        }
+    } else {
+        if (result.size() >= ks->offset) {
+            ks->setKey(KeyRef(ks->arena(), result[ks->offset - 1].key));
+            ks->offset = 1;
+        } else {
+            ks->setKey(KeyRef(ks->arena(), keyAfter(result[result.size()-1].key)));
+            ks->offset -= result.size();
+        }
+    }
+    TraceEvent("NormalizeKeySelector").
+        detail("NormalizedKey", ks->getKey()).
+        detail("NormalizedOffset", ks->offset).
+        detail("PrivateKeyRangeStart", range.begin).
+        detail("PrivateKeyRangeEnd", range.end);
+    return Void();
+}
+
+ACTOR Future<Standalone<RangeResultRef>> getRangeAggregationActor(
+    PrivateKeySpace* pks,
+    Reference<ReadYourWritesTransaction> ryw,
+    KeySelector begin,
+    KeySelector end,
+    GetRangeLimits limits,
+    bool reverse )
+{   
+    // This function handles ranges which cover more than one keyrange and aggregates all results
+    // KeySelector, GetRangeLimits and reverse are all handled here
+    
+    // make sure orEqual == false
+    if(begin.orEqual)
+		begin.removeOrEqual(begin.arena());
+    if(end.orEqual)
+        end.removeOrEqual(end.arena());
+
+    // make sure offset == 1
+    state RangeMap<Key, PrivateKeyRangeBaseImpl*, KeyRangeRef>::Iterator iter =
+        pks->getKeyRangeMap().rangeContaining(begin.getKey());
+    while (begin.offset != 1 && iter != pks->getKeyRangeMap().ranges().begin()) {
```

Comment:
check also the ranges().end() for jumping out the loop.

Reply:
Done.  Thanks.